File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def eat(speed):
            hour = 0
            for p in piles:
                hour += p//speed
                if p%speed:
                    hour += 1

            return hour

        l, r = 1, max(piles)
        ans = r

        while l<=r:
            mid = (l+r)//2

            hours = eat(mid)
            if hours<=h:
                ans = min(ans, mid)
                r = mid - 1
            else:
                l = mid + 1
        

        return ans

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.c
int minEatingSpeed(int* piles, int pilesSize, int h) {
    int slow=1, fast=0;

    // find the largest pile
    for (int i=0; i<pilesSize; ++i) {
        if (piles[i] > fast) {
            fast = piles[i];
        }
    }
    int ans = fast;
    while (slow <= fast) {
        int speed = (slow + fast) / 2;
        unsigned int hour = 0;
        for (int i=0; i<pilesSize; ++i) {
            hour += piles[i] / speed;
            if (piles[i] % speed != 0)
                hour++;
        }

        if (hour <= h) {
            ans = fmin(ans, speed);
            fast = speed - 1;
        }
        else {
            slow = speed + 1;
        }
    }

    return ans;
}